frontend/lib/resolution/Resolver.cpp

```diff
@@ -1933,110 +1940,153 @@ void Resolver::resolveTupleUnpackDecl(const TupleDecl* lhsTuple,
       CHPL_REPORT(context, TupleDeclNotTuple, lhsTuple, rhsType.type());
       return;
     }
+
+  // Then, check that they have the same size
   } else if (lhsTuple->numDecls() != rhsT->numElements()) {
-    // Then, check that they have the same size
     CHPL_REPORT(context, TupleDeclMismatchedElems, lhsTuple, rhsT);
     return;
+
+  // Else, it's a tuple of the same size, so use the RHS element types
   } else {
+    QualifiedType::Kind lhsKind = (Qualifier) lhsTuple->intentOrKind();
+    bool isLhsIdxOrRef = lhsKind == Qualifier::INDEX ||
+                         isRefQualifier(lhsKind);
+
     for (int i = 0; i < rhsT->numElements(); i++) {
-      eltTypes.push_back(rhsT->elementType(i));
+      auto kind = rhsT->elementType(i).kind();
+      auto type = rhsT->elementType(i).type();
+
+      // If the LHS is a ref tuple or a tuple index e.g., for (i, j)..., and
+      // the RHS is a ref tuple, then we need to transform the detupled
+      // components into references to the original tuple's components.
+      if (isLhsIdxOrRef && rhsType.isRef()) kind = rhsType.kind();
+      eltTypes.push_back({ kind, type });
     }
   }
 
   // Finally, try to resolve the types of the elements
   int i = 0;
   for (auto actual : lhsTuple->decls()) {
     QualifiedType rhsEltType = eltTypes[i];
+
     if (auto innerTuple = actual->toTupleDecl()) {
       resolveTupleUnpackDecl(innerTuple, rhsEltType);
+
     } else if (auto namedDecl = actual->toNamedDecl()) {
       resolveNamedDecl(namedDecl, rhsEltType.type());
+
+      // If the element is an index variable, then use the RHS's intent.
```

Comment:
Not clear to me why this is needed, could you explain?

Reply:
If I don't do this then the component will have the `INDEX` intent, which isn't concrete.

Reply:
But why do you make it `const`?

Reply:
Ah, I see your comment below.